File: solutions/fsharp/binary-search/1/BinarySearch.fs
module BinarySearch

let find (input: int[]) (value: int) = 
    let rec search low high =
        if low > high then
            None
        else
            let mid = (low + high) / 2
            if input.[mid] = value then
                Some mid
            elif input.[mid] > value then
                search low (mid - 1)
            else
                search (mid + 1) high
    search 0 (Array.length input - 1)

File: solutions/ocaml/binary-search/1/binary_search.ml
let find arr target =
  let rec binary_search arr target low high =
    if low > high then
      Error "value not in array"
    else
      let mid = (low + high) / 2 in
      let mid_element = arr.(mid) in
      if mid_element = target then
        Ok mid
      else if mid_element > target then
        binary_search arr target low (mid - 1)
      else
        binary_search arr target (mid + 1) high
  in
  binary_search arr target 0 (Array.length arr - 1)
